Emulator backend API support

keith-zephyr opened the pull request:
Add a mechanism for emulators of the same device type to declare a common backend API. 

See [RFC #53791](https://github.com/zephyrproject-rtos/zephyr/issues/53791) for more information.

keith-zephyr committed:
emul: Add support for backend APIs

The emulator structure currently initializes a bus-level API that is
used to communicated register accesses to the emulator from the parent
bus.

Add support for declaring a backend API for emulators. This provides
tests a common way to interact with emulators of the same device type.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
test: emul: Verify backend API operation

Verify backend API is functional with the emulator subsystem.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
emul: add stub driver for emulators

Not all emulators correspond to a real driver.  Notably some tests don't
require a real driver.  Provide a macro that initializes a stub driver.

Signed-off-by: Keith Short <keithshort@google.com>

alevkoy commented on include/zephyr/drivers/emul_stub_device.h:
```diff
@@ -0,0 +1,42 @@
+		return 0;                                                                          \
+	}                                                                                          \
+                                                                                                   \
+	/* Since this is only stub, allocate the structs once. */                                  \
```
Would they be allocated more times if this wasn't a stub?

alevkoy replied:
Are you going to get rid of it?

alevkoy commented on include/zephyr/drivers/emul.h:
```diff
@@ -120,8 +122,9 @@ extern const struct emul __emul_list_end[];
+ * @param backend_api emulator-specific backend api
  */
-#define EMUL_DT_DEFINE(node_id, init_fn, data_ptr, cfg_ptr, bus_api)                               \
+#define EMUL_DT_DEFINE(node_id, init_fn, data_ptr, cfg_ptr, bus_api, _backend_api)                 \
```
Why the leading underscore?

alevkoy commented on tests/subsys/emul/src/emul_test_driver.c:
```diff
@@ -0,0 +1,82 @@
+	return 0;
+}
+
+#define EMUL_TEST_DRIVER(n)                                                                        \
```
I can't tell whether this is a structural issue or just semantics, but AFAICT, the point of the stub driver is to satisfy the requirement for a linked `struct device` for `struct emul`s that don't have a natural one. If that's true, then it seems to me like the names of this module and its members should just be "emul_something" and not "something_driver." The test is written in terms of the backend API, which is an emulator API. From the point of view of other modules, this module provides an emulator, and the fact that it contains a driver is an implementation detail.

aaronemassey reviewed (changes requested):
I quite like this for making it easier to write API specific tests instead of chip/driver specific tests. The API can then easily be tested with various drivers/chips but the test can remain the same.

alevkoy reviewed (commented):
Looks good except for the comment that still doesn't make sense.

keith-zephyr commented:
Note - commit https://github.com/zephyrproject-rtos/zephyr/commit/774949c7346a0e3984beb8aa485741f08fcb3f96 is a rebase only.

keith-zephyr commented:
> Looks good except for the comment that still doesn't make sense.

Fixed.

keith-zephyr commented:
@yperess Can you give this a review?

yperess reviewed (changes requested):
Please add some details of this to https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/hardware/emulator/index.rst. I think it'll be interesting to show an example like we discussed offline where we want to run the same test using 2 different drivers of a similar "class" and allowing the same test to exercise both drivers.

keith-zephyr committed:
docs: Update emulator documentation

Update emulator documentation with a diagram and description for how to
use the .bus_api and ._backend_api parameters with EMUL_DT_DEFINE().

Signed-off-by: Keith Short <keithshort@google.com>
